# I. Thực hành

Câu 1: Trong phương pháp bag-of-word thì mỗi một đoạn văn bản sẽ được biến đổi thành véc tơ đặc trưng như thế nào?


A. Thành véc tơ tần suất của các từ xuất hiện trong từ điển.

B. Thành véc tơ one-hot, từ nào xuất hiện trong văn bản thì có giá trị 1, trái lại nhận giá trị 0.

C. Thành véc tơ tf-idf của từ.

D. Thành một véc tơ trong không gian latent space.

A. Thành véc tơ tần suất của các từ xuất hiện trong từ điển.

Câu 2: Hai phương pháp `bigram` và `trigram` trong `bag-of-n-gram` sẽ mã hoá một văn bản như thế nào?

A. `Bigram` sẽ ghép ba từ liên tiếp thành một token. `Trigram` sẽ ghép hai từ liên tiếp nhau thành một token.

B. `Bigram` sẽ ghép hai từ liên tiếp thành một token. `Trigram` sẽ ghép ba từ liên tiếp nhau thành một token.

C. `Bigram` sẽ chia mỗi một từ thành một token. `Trigram` sẽ ghép ba từ liên tiếp nhau thành một token.

D. `Bigram` sẽ ghép hai từ bất kì trong khoảng cách gần thành một token. `Trigram` sẽ ghép ba từ liên tiếp trong khoảng cách gần thành một token.

B. Bigram sẽ ghép hai từ liên tiếp thành một token. Trigram sẽ ghép ba từ liên tiếp nhau thành một token.

Câu 3: Giải thích ý nghĩa của chỉ số tf-idf được sử dụng để mã hoá các từ trong bộ văn bản. Một từ có tf-idf cao thì chứng tỏ điều gì?

A. `tf-idf` là chỉ số mã hóa dựa trên độ dài của văn bản. Khi `tf-idf` cao chứng tỏ mức độ phổ biến của từ càng cao và ngược lại.

B. `tf-idf` là chỉ số mã hóa dựa trên tần suất. Khi `tf-idf` cao chứng tỏ mức độ phổ biến của từ càng thấp và ngược lại.

C. `tf-idf` nhằm đánh giá mức độ phổ biến của một từ dựa vào tần suất xuất hiện của từ trong một văn bản và tần suất các văn bản xuất hiện từ đó. Khi `tf-idf` cao chứng tỏ mức độ phổ biến của từ càng cao và ngược lại.

D. `tf-idf` nhằm đánh giá mức độ quan trọng của một từ dựa vào tần suất xuất hiện của từ trong một văn bản và tần suất các văn bản xuất hiện từ đó. Khi `tf-idf` cao chứng tỏ mức độ quan trọng của từ càng cao và ngược lại.

D. tf-idf nhằm đánh giá mức độ quan trọng của một từ dựa vào tần suất xuất hiện của từ trong một văn bản và tần suất các văn bản xuất hiện từ đó. Khi tf-idf cao chứng tỏ mức độ quan trọng của từ càng cao và ngược lại.

Câu 4: Kiến trúc nào thường được sử dụng để trích xuất đặc trưng (`Feature Extractor`) trong các bộ dữ liệu ảnh:

A. Thuật toán SVM
 
B. Thuật toán MLP

C. Thuật toán CNN

D. Thuật toán LSTM.

C. Thuật toán CNN

Câu 5: Tại sao chúng ta cần thực hiện các kĩ thuật chuẩn hóa dữ liệu như `Min-max Scaling`, `Standardization Scaling`, hoặc `Unit length Scaling` trên dữ liệu $\mathbf{X}$ trước khi huấn luyện mô hình:

A. Để đồng nhất đơn vị của các biến trước khi huấn luyện mô hình.

B. Giúp các thuật toán ổn định hơn.

C. Giảm chi phí tính toán.

D. Cả ba đáp án trên.

D. Cả ba đáp án trên.

# II. Thực hành

Câu 6: Thực hành phân loại văn bản dựa trên phương pháp bag-of-word và tf-idf đối với bộ dữ liệu [10Topics](https://github.com/duyvuleo/VNTC/tree/master/Data/10Topics/Ver1.1).

Gợi ý: Bạn cần thực hiện theo các bước sau đây:

- Bước 1: Download dữ liệu.

- Bước 2: Tiền xử lý dữ liệu (`Data pre-processing`). Sử dụng phương pháp `bag-of-word` để mã hóa văn bản thành véc tơ.

- Bước 3: Phân chia tập `train/validation` và huấn luyện mô hình.

**Download Data**

In [6]:
# Google drive mount point
#from google.colab import drive
#drive.mount('/content/drive')

In [ ]:
#import os
#os.chdir('drive/MyDrive/Colab Notebooks/ML-Handson-With-Python2')

In [23]:
import os
os.getcwd()

'C:\\Users\\nguye\\Documents\\ML HAndson'

In [4]:
!wget Test_Full.rar \
&& unrar x Test_Full.rar \
&& rm -f Test_Full.rar && ls Test_Full

'wget' is not recognized as an internal or external command,
operable program or batch file.


In [19]:
import numpy as np
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.compose import ColumnTransformer
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import LinearSVC
from sklearn.neural_network import MLPClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.impute import KNNImputer, SimpleImputer
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.model_selection import cross_val_score
from sklearn.metrics import fbeta_score
from sklearn.metrics import make_scorer
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.preprocessing import OneHotEncoder, MinMaxScaler

**Data preprocessing**

**Model training**

Câu 7: Thực hành biến đổi dữ liệu với python

- Lấy ra current date và current time
- Current year
- Month of year
- Week number of the year
- Weekday of the week
- Day of year
- Day of the month
- Day of week


Gợi ý: Sử dụng package datetime.

Câu 8: Sử dụng AutoEncoder để giảm chiều bộ dữ liệu sau từ 50 chiều về 10 chiều.
Code khởi tạo dữ liệu:

```
from sklearn.datasets import make_classification

# Khởi tạo dữ liệu example
X, y = make_classification(n_samples=500, n_features=50, random_state=123)
```

Câu 9: Xây dựng mô hình phân loại trên 10 biến trong không gian giảm chiều và đánh giá độ chính xác mô hình.

Câu 10: Sử dụng các kĩ thuật lựa chọn đặc trưng khác nhau để lựa chọn ra 10 biến đầu vào từ 50 biến đầu vào gốc. Xây dựng mô hình phân loại trên các biến được lựa chọn và đánh giá độ chính xác.

In [84]:
# Example 1
f = open('01_01_2020.txt', encoding='utf8').read()
f

'Main index falls at year-end session\n\nThe HCMC bourse faced gloomy trade in the last session of 2019, witnessing the VN-Index fall 4.04 points to end the year at 960.99 on December 31 due to heavy sell-offs and cautious investor sentiment.\n\nOn the HCMC market, market breadth was negative as losers outnumbered gainers by 199 to 137. The main index lost 0.42% against the session earlier. There were 182.24 million shares worth VND3 trillion changing hands, up 4.2% in volume, but down 1.6% in value versus the day earlier. Block deals accounted for more than 40 million shares valued at VND859 billion.\n\nConstruction firm ROS continued its downward spiral, closing the session down 6.9%, piling pressure on the southern exchange.\n\nSome bank stocks, such as VCB, STB, MBB and CTG, lost steam, while others picked up slightly and lender HDB soared by 2.4% to close at the intraday high.\n\nAmong blue-chip stocks, property firm NVL and carrier VJC added over 1%, while real estate firm VHM, b

In [85]:
import pandas as pd
import numpy as np

import nltk

from nltk.tokenize import TweetTokenizer
tknzr = TweetTokenizer(strip_handles=True)

from nltk.stem import WordNetLemmatizer
wnl = WordNetLemmatizer()

from nltk.corpus import sentiwordnet as swn
f_tokenize = tknzr.tokenize(f.lower())
f_stem = []
for element in f_tokenize:
    f_stem.append(wnl.lemmatize(element))
len(f_stem)

454

In [122]:
f_stem_no_integers = [item for item in f_stem if not item.isdigit()]
len(f_stem_no_integers) #bỏ những ký tự số

445

In [119]:
from sklearn.feature_extraction.text import TfidfVectorizer
# Tính tfidf cho mỗi từ. max_df để loại bỏ stopwords xuất hiện ở hơn 90% các câu
vectorizer = TfidfVectorizer()
# Tokenize các câu theo tfidf
X = vectorizer.fit_transform([f])
#print(vectorizer.get_feature_names())
X

<1x218 sparse matrix of type '<class 'numpy.float64'>'
	with 218 stored elements in Compressed Sparse Row format>

In [96]:
df = pd.DataFrame(X.T.todense(), index=vectorizer.get_feature_names(), columns=["tfidf"]) 
df.sort_values(by=["tfidf"],ascending=False)
df

tfidf
04          0.022559
06          0.022559
102         0.022559
137         0.022559
182         0.022559
...              ...
while       0.067677
with        0.022559
witnessing  0.022559
worth       0.090236
year        0.067677

[218 rows x 1 columns]

In [115]:
df.loc['year'].values[0]

0.06767682521924123

In [109]:
'year' in df.index.tolist()

True

In [116]:
def sentiment_score_pos_tag (f_stem, df):
    pos_score = neg_score = token_count = obj_score = 0
    for word in f_stem:
        tag = nltk.pos_tag([word])
        synst = None
        if 'NN' in tag[0] and list(swn.senti_synsets(word, 'n')):
            synst = list(swn.senti_synsets(word, 'n'))[0]
        elif 'VB' in tag[0] and list(swn.senti_synsets(word, 'v')):
            synst = list(swn.senti_synsets(word, 'v'))[0]
        elif 'JJ' in tag[0] and list(swn.senti_synsets(word, 'a')):
            synst = list(swn.senti_synsets(word, 'a'))[0]
        elif 'RB' in tag[0] and list(swn.senti_synsets(word, 'r')):
            synst = list(swn.senti_synsets(word, 'r'))[0]
          
        if synst:
            token_count += 1
            if word in df.index.tolist():
                pos_score += synst.pos_score()* df.loc[word].values[0]
                neg_score += synst.neg_score()* df.loc[word].values[0]
                obj_score += synst.obj_score()* df.loc[word].values[0]
                
            else:
                pos_score += synst.pos_score()
                neg_score += synst.neg_score()
                obj_score += synst.obj_score()
                
        
    score_dict = {'positive_score': pos_score, 'negative_score': neg_score,
                       'objectivity_score': obj_score,'token_count': token_count}
    return score_dict 

Tính sentiment scores của 1 bài báo bằng tổng (sentiment score của từng token * tfidf của từng token)

In [123]:
sentiment_score_pos_tag (f_stem_no_integers,df)

{'positive_score': 1.823316560654178,
 'negative_score': 0.5723907402227868,
 'objectivity_score': 32.23696082199501,
 'token_count': 161}

Điểm số nếu không tính tdidf thể hiện tốt hơn xu hướng của bài viết: negative score > positive score 
{'positive_score': 4.25, 'negative_score': 5.375, 'objectivity_score': 151.375, 'token_count': 161}

In [39]:
economy = pd.read_table('economy_daily_sentiment_score', sep=',')
economy

publish_date                                          file_name  \
0     2020-01-01  01_01_202001_01_2020_101_01_2020_201_01_2020_3...   
1     2020-01-02     01_02_202001_02_2020_101_02_2020_201_02_2020_3   
2     2020-01-03  01_03_202001_03_2020_101_03_2020_201_03_2020_3...   
3     2020-01-04  01_04_202001_04_2020_101_04_2020_201_04_2020_3...   
4     2020-01-05                                         01_05_2020   
..           ...                                                ...   
542   2021-06-26                 06_26_202106_26_2021_106_26_2021_2   
543   2021-06-27                                         06_27_2021   
544   2021-06-28     06_28_202106_28_2021_106_28_2021_206_28_2021_3   
545   2021-06-29  06_29_202106_29_2021_106_29_2021_206_29_2021_3...   
546   2021-06-30  06_30_202106_30_2021_106_30_2021_206_30_2021_3...   

     positive_score  negative_score  objectivity_score  token_count  
0            96.500          65.375           2311.125       2473.0  
1           114.875          63.125           1700.000       1878.0  
2            51.125          39.500            921.375       1012.0  
3            61.500          37.625           1387.875       1487.0  
4               NaN             NaN                NaN          NaN  
..              ...             ...                ...          ...  
542          50.125          25.750            910.125        986.0  
543           3.375           3.375            100.250        107.0  
544          37.625          26.375            725.000        789.0  
545          73.875          46.250           1662.875       1783.0  
546          26.125          19.250            658.625        704.0  

[547 rows x 6 columns]

In [40]:
from datetime import datetime
def parser(x):
    return datetime.strptime(x, '%Y-%m-%d')
economy['publish_date'] = economy['publish_date'].map(lambda x: parser(x))

In [41]:
economy['weekday'] = economy['publish_date'].apply(lambda x: x.date().strftime('%A'))
economy['weekday']

0      Wednesday
1       Thursday
2         Friday
3       Saturday
4         Sunday
         ...    
542     Saturday
543       Sunday
544       Monday
545      Tuesday
546    Wednesday
Name: weekday, Length: 547, dtype: object

In [42]:
economy['year_week_weekday'] = economy['publish_date'].apply(lambda x: x.date().isocalendar())
economy['year_week_weekday']

0       (2020, 1, 3)
1       (2020, 1, 4)
2       (2020, 1, 5)
3       (2020, 1, 6)
4       (2020, 1, 7)
           ...      
542    (2021, 25, 6)
543    (2021, 25, 7)
544    (2021, 26, 1)
545    (2021, 26, 2)
546    (2021, 26, 3)
Name: year_week_weekday, Length: 547, dtype: object

In [43]:
economy

publish_date                                          file_name  \
0     2020-01-01  01_01_202001_01_2020_101_01_2020_201_01_2020_3...   
1     2020-01-02     01_02_202001_02_2020_101_02_2020_201_02_2020_3   
2     2020-01-03  01_03_202001_03_2020_101_03_2020_201_03_2020_3...   
3     2020-01-04  01_04_202001_04_2020_101_04_2020_201_04_2020_3...   
4     2020-01-05                                         01_05_2020   
..           ...                                                ...   
542   2021-06-26                 06_26_202106_26_2021_106_26_2021_2   
543   2021-06-27                                         06_27_2021   
544   2021-06-28     06_28_202106_28_2021_106_28_2021_206_28_2021_3   
545   2021-06-29  06_29_202106_29_2021_106_29_2021_206_29_2021_3...   
546   2021-06-30  06_30_202106_30_2021_106_30_2021_206_30_2021_3...   

     positive_score  negative_score  objectivity_score  token_count  \
0            96.500          65.375           2311.125       2473.0   
1           114.875          63.125           1700.000       1878.0   
2            51.125          39.500            921.375       1012.0   
3            61.500          37.625           1387.875       1487.0   
4               NaN             NaN                NaN          NaN   
..              ...             ...                ...          ...   
542          50.125          25.750            910.125        986.0   
543           3.375           3.375            100.250        107.0   
544          37.625          26.375            725.000        789.0   
545          73.875          46.250           1662.875       1783.0   
546          26.125          19.250            658.625        704.0   

       weekday year_week_weekday  
0    Wednesday      (2020, 1, 3)  
1     Thursday      (2020, 1, 4)  
2       Friday      (2020, 1, 5)  
3     Saturday      (2020, 1, 6)  
4       Sunday      (2020, 1, 7)  
..         ...               ...  
542   Saturday     (2021, 25, 6)  
543     Sunday     (2021, 25, 7)  
544     Monday     (2021, 26, 1)  
545    Tuesday     (2021, 26, 2)  
546  Wednesday     (2021, 26, 3)  

[547 rows x 8 columns]